improve unique performance on CPU

mingfeima opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107846

Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

mingfeima committed:
improve unique performance on CPU

[ghstack-poisoned]

mingfeima committed:
Update on "improve unique performance on CPU"




[ghstack-poisoned]

mingfeima committed:
Update on "improve unique performance on CPU"




[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+
+  if (return_counts) {
+    // do diff to get count
+    at::parallel_for(0, unique_count, 0, [&](int64_t begin, int64_t end) {
+      for (const auto i : c10::irange(begin, end)) {
+        counts_data[i] = unique_index_data[i + 1] - unique_index_data[i];
+      }
+    });
```
This looks like `unique_index[1:] - unique_index[:-1]`? So it appears to be just a difference of slices and we can use high level functions here.

mingfeima replied:
Yes, we can do a `torch.diff` here. But like the previously one, performance would be a little bit slower, roughly add another ~5% overhead.

nikitaved commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+  // the first element is always true
+  mask_acc[0] = true;
+
+  // we can parallel on [1, numel) but we need to make sure it has
+  // the same parallel scope with the next loop
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    if (begin == 0) { begin += 1; }
+    for (const auto i : c10::irange(begin, end)) {
+      mask_acc[i] = input_sorted_data[i] != input_sorted_data[i - 1];
+    }
+  });
```
`mask_acc = input_sorted_data[1:] != input_sorted_data[:-1]`. These could be accessed with either `narrow` or `slice`.

mingfeima replied:
Yes, we can do it with `narrow`. But from the performance perspective, I personally preferred directly manipulating data pointers. Using pointers is always the fastest and `TensorAccessor` is almost as fast as pointers.

Firstly, `narrow` or `slice` will create a tensor view which allocates a new `TensorImpl`.
Also, `eq` is implemented by tensor iterator and this also takes some extra overhead; and usually the dispatched functions have quite a long calling stack.
Additionally, if we insert `eq`, its parallel pattern may be different from the following loops (data movement between cpu cores)


I modified this part to the following:
```C++
auto input_sorted_0 = input_sorted.narrow(0, 0, numel - 1);
auto input_sorted_1 = input_sorted.narrow(0, 1, numel - 1);
auto mask_1 = mask.narrow(0, 1, numel - 1);
at::eq_out(input_sorted_0, input_sorted_1, mask_1);
```

The performance the benchmark changed from
```BASH
# test dtype of  torch.int64
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```
to
```BASH
# test dtype of  torch.int64
Torch just sort: 0.29884862899780273 s
Torch sort + indexes: 0.327167272567749 s
```

nikitaved commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+    }
+  });
+
+  // handle equal the NaN's
+  bool last_element_isnan = _isnan<scalar_t>(input_sorted_data[numel - 1]);
+  if (input.is_floating_point() && equal_nan && last_element_isnan) {
+    int64_t firstnan_index = numel;
+    while (_isnan<scalar_t>(input_sorted_data[firstnan_index - 1])) {
+      firstnan_index--;
+    }
+    mask_acc[firstnan_index] = true;
+    if (firstnan_index + 1 < numel) {
+      for (const auto i : c10::irange(firstnan_index + 1, numel)) {
+        mask_acc[i] = false;
+      }
+    }
+  }
```
This pattern looks like a lower bound with a custom comparator?

mingfeima replied:
`std::find_if` fits in better here

nikitaved commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+    }
+  });
+
+  // calculate total count of unique and thread offset in output
+  int64_t unique_count = std::accumulate(unique_count_thread.begin(), unique_count_thread.end(), 0);
+  int64_t sum = 0;
+  for (const auto t : c10::irange(num_threads)) {
+    offset_thread[t] = sum;
+    sum += unique_count_thread[t];
+  }
```
nit: seems like `sum` and `unique_count` in the end are the same? So, I guess, these two loops could be merged...

mingfeima replied:
Oh yes, thanks for the remainder !

jgong5 commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+// `inverse_indices` and `counts` computation in a fused loop,
+// which made this part almost a free launch.
+//
+// This kernel also implements a `equal_nan` flag which has same
+// function as NumPy's unique. Currently this is always disabled.
```
What's the behavior of CUDA implementation? Are CPU and CUDA implementation aligned? There is an open issue https://github.com/pytorch/pytorch/issues/94106 about the issue with nan handling. Can this PR address that problem?

mingfeima replied:
right now `torch.unique` treates NaN as different values (both CPU and CUDA), this behavior is different from numpy.unique which has a `equal_nan` argument. This patch won't change the behavior of `torch.unique`, but I also implemented `equal_nan` in the kernel so if someone decides that this feature shall be enabled on torch, it can be done very easily.

jgong5 commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+  auto input_flattened = input.flatten();
+
+  Tensor input_sorted, indices;
+  std::tie(input_sorted, indices) = input_flattened.sort();
```
Are you going to factor out the implementation for `torch.unique_consecutive`? This function is a combination of `sort` and `torch.unique_consecutive`. I think the optimization here can also benefit unique_consecutive, right?

mingfeima replied:
will have it done later on, right now `torch.unique_consecutive` is sequential.

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

nikitaved commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+    auto firstnan = std::find_if(input_sorted_data, input_sorted_data + numel, is_nan);
+    int64_t firstnan_index = std::distance(input_sorted_data, firstnan);
+
+    mask_acc[firstnan_index] = true;
+    if (firstnan_index + 1 < numel) {
+      for (const auto i : c10::irange(firstnan_index + 1, numel)) {
+        mask_acc[i] = false;
+      }
+    }
+  }
```
I would still wrap with block in parallel_for in an unlikely even of having a significant portion of nans.

mingfeima replied:
Oh, good idea !

mingfeima replied:
updated!

nikitaved commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +145,159 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+    }
+  }
+
+  // calculate unique count from each thread
+  at::parallel_for(0, numel, 0, [&](int64_t begin, int64_t end) {
+    int tid = at::get_thread_num();
```
It is always beneficial to set grain_size to zero to force parallelization?

mingfeima replied:
No. For a small problem size, parallel it would be slower. But in practice, the difference of setting grain size or not does not matter very much. Because if the problem size is very small, then it would only take very small amount of time.

Usually for simple operators we set grain_size to be `at::internal::grain_size` which is 32K and for complex operators just set grain_size to 0. And the parallel loop in this patch has been changed to use the default grain size.

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

mingfeima commented:
@nikitaved could you please help review this one again ? thanks!

nikitaved reviewed (approved):
LGTM! Thank you, @mingfeima !

mingfeima commented:
@pytorchbot merge

mingfeima commented:
@nkaretnikov could you please help review this one ?

nkaretnikov commented:
@mingfeima I'm not actively working on PyTorch at the moment. But I'll try to find someone to review this.

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +146,164 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+  at::parallel_for(0, numel, grain_size, [&](int64_t begin, int64_t end) {
+    if (begin == 0) { begin += 1; }
+    for (const auto i : c10::irange(begin, end)) {
+      mask_acc[i] = input_sorted_data[i] != input_sorted_data[i - 1];
```
This is a good idea! from the given benchmark, using `re-computing` is only slightly faster than `mask`, but i suppose the benefit will be bigger if we have a larger input tensor.

```shell
### this is data with mask, the old implementation
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s

### this is data with re-computing, the current implementation
replace masking:
Torch just sort: 3.477884292602539 s
Torch sort + indexes: 3.547271966934204 s
```

mingfeima commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -143,6 +146,164 @@ std::tuple<Tensor, Tensor, Tensor> unique_cpu_template(
+  });
+
+  // handle equal the NaN's
+  bool last_element_isnan = _isnan<scalar_t>(input_sorted_data[numel - 1]);
```
updated, `equal_nan`  is added in template so as to skip unnecesary check

mingfeima commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -318,20 +479,27 @@ std::tuple<Tensor, Tensor, Tensor> _unique_dim_cpu_template(
+    if (sorted && self.scalar_type() != kBool) {
+      std::tie(output, inverse, std::ignore) = unique_cpu_sorted_template<scalar_t>(self, return_inverse, false);
+    } else {
+      std::tie(output, inverse, std::ignore) = unique_cpu_template<scalar_t>(self, sorted, return_inverse, false);
```
Good idea ! Now the `unique_elements` specification with bool is sequential, makes more sense to do a parallel reduce, will take care of this later on.

peterbell10 replied:
I don't understand why we need a hashmap based solution? Just ignore the `sorted` argument and always return sorted output. That's what CUDA does.

mingfeima replied:
So you are suggesting to do sorting no matter whether the `sorted` argument is true or not ? Actually that would be easier.

mingfeima replied:
Updates: the hash table implementation is removed.
TODO: add `bool` specification to match UniqueCub.cu

mingfeima replied:
@peterbell10 could you please review again, this patch has been updated!

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/Unique.cpp:
```diff
@@ -192,6 +75,8 @@ struct IsUnique<scalar_t, true> {
 // This kernel also implements a `equal_nan` flag which has same
 // function as NumPy's unique. Currently this is always disabled.
 //
+// TODO: add `bool` specification, use similar approach as UniqueCub
```
NIT: 
```suggestion
// TODO: add `bool` specialization, use similar approach as UniqueCub
```

mingfeima committed:
Update on "improve unique performance on CPU"


Fix https://github.com/pytorch/pytorch/issues/107098, improve `unique` performance on CPU.

The algorithm is taken from Numpy implementation at https://github.com/numpy/numpy/blob/main/numpy/lib/arraysetops.py#L323, it first do a sort on the input sequence and then  use a `mask` to record the unique element of each consecutive section.

Now we don't have parallel sort on 1-dimension float tensor, will have it enabled in next step. Parallel radix sort is used for 1-dimensional int tensor.

The following data is collected with script in the issue on Intel(R) Xeon(R) Gold 6248 CPU @ 2.5GHz with single sockets (20 cores):

#### before (dtype int64)
```
Numpy just sort: 0.4271528720855713 s
Numpy sort + indexes: 6.383563041687012 s
Torch just sort: 0.46924352645874023 s
Torch sort + indexes: 1.8140404224395752 s
```

#### after (dtype int64)
```
Torch just sort: 0.2540090084075928 s
Torch sort + indexes: 0.2766146659851074 s
```

#### before (float32)
```
Numpy just sort: 0.41129398345947266 s
Numpy sort + indexes: 6.422696590423584 s
Torch just sort: 9.109549283981323 s
Torch sort + indexes: 37.59021711349487 s
```

#### after (float32)
```
Torch just sort: 3.5369982719421387 s
Torch sort + indexes: 3.582240581512451 s
```

if we enabled parallel sort on 1-dimension float tensor, the performance is:
```
Torch just sort: 0.3212606906890869 s
Torch sort + indexes: 0.36211371421813965 s
```

Since i have fused the `inverse_indices` and `count` calculation in fused parallel loop (the algorithm is identical to NumPy's but with better optimization), they will take a small amount of additional time.

[ghstack-poisoned]

mingfeima commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "Caused Meta-internal issues S369412" -c ghfirst

kit1980 commented:
@mingfeima I can provide some details later if you want to reland this.

malfet commented:
@mingfeima your PR is not a culprit, feel free to reland after i close https://github.com/pytorch/pytorch/issues/111189 (i.e. the memory corruption comes from the `torch.sort` which is due to a bug in `fbgemm::radix_search`
